librepo/metadata_downloader.c

```diff
@@ -0,0 +1,410 @@
+
+    target->handle = handle;
+    target->repo = repo;
+    target->repomd = repomd;
+    target->cbdata = cbdata;
+    target->repomd_records_to_download = 0;
+    target->repomd_records_downloaded = 0;
+
+    return target;
+}
+
+LrMetadataTarget *
+lr_metadatatarget_new2(LrHandle *handle,
+                       void *cbdata,
+                       LrProgressCb progresscb,
+                       LrMirrorFailureCb mirrorfailure_cb,
+                       LrEndCb endcb,
+                       GError **err)
+{
+    LrMetadataTarget *target = lr_metadatatarget_new(handle, NULL, NULL, cbdata, err);
+    target->progresscb = progresscb;
+    target->mirrorfailurecb = mirrorfailure_cb;
+    target->endcb = endcb;
+
+    return target;
+}
+
+void
+lr_metadatatarget_free(LrMetadataTarget *target)
+{
+    if (!target)
+        return;
+    g_string_chunk_free(target->chunk);
+    g_free(target);
+}
+
+static gboolean
+lr_setup_sigaction(struct sigaction *old_sigact,
+                   gboolean interruptible,
+                   GError **err)
+{
+    // Setup sighandler
+    if (interruptible) {
+        struct sigaction sigact;
+        g_debug("%s: Using own SIGINT handler", __func__);
+        memset(&sigact, 0, sizeof((*old_sigact)));
+        memset(&sigact, 0, sizeof(sigact));
+        sigemptyset(&sigact.sa_mask);
+        sigact.sa_handler = lr_sigint_handler;
+        sigaddset(&sigact.sa_mask, SIGINT);
+        sigact.sa_flags = SA_RESTART;
+        if (sigaction(SIGINT, &sigact, old_sigact) == -1) {
+            g_set_error(err, LR_PACKAGE_DOWNLOADER_ERROR, LRE_SIGACTION,
+                        "Cannot set Librepo SIGINT handler");
+            return FALSE;
+        }
+    }
+    return TRUE;
+}
+
+static gboolean
+lr_restore_sigaction(struct sigaction *old_sigact,
+                     gboolean interruptible,
+                     GError **err)
+{
+    // Restore original signal handler
+    if (interruptible) {
+        g_debug("%s: Restoring an old SIGINT handler", __func__);
+        sigaction(SIGINT, old_sigact, NULL);
+        if (lr_interrupt) {
+            if (err && *err != NULL)
+                g_clear_error(err);
+            g_set_error(err, LR_PACKAGE_DOWNLOADER_ERROR, LRE_INTERRUPTED,
+                        "Insterupted by a SIGINT signal");
+            return FALSE;
+        }
+    }
+
+    return TRUE;
+}
+
+static gboolean
+lr_metadata_download_cleanup(GSList *download_targets,
+                             GError **err)
+{
+    gboolean ret = TRUE;
+
+    // Copy download statuses from download_targets to targets
+    for (GSList *elem = download_targets; elem; elem = g_slist_next(elem)) {
+        LrDownloadTarget *download_target = elem->data;
+        LrMetadataTarget *target = download_target->userdata;
+        if (download_target->err)
+            target->err = g_string_chunk_insert(target->chunk,
+                                                download_target->err);
```

Comment:
Hm, so if say the repodata is out of sync where we get a new `repomd.xml` but old versions of other things, this will end up concatenating together all of the errors?